File: web/src/keyenc.ts
/**
 * keyenc — join several untrusted strings into one key that no component's
 * content can forge.
 *
 * The TypeScript half of github.com/cplieger/keyenc. It produces
 * byte-identical keys to the Go half for every valid-UTF-8 input, pinned by the
 * shared conformance fixture in `conformance/keys.json`. That parity is
 * load-bearing wherever a key built in one language is looked up against keys
 * built in the other.
 *
 * See the Go package documentation for the grammar, the nesting rule and the
 * reasoning behind the size bound. The short version: `:` separates components
 * and `\` escapes, both escaped element-wise inside every component, so a
 * component carrying a separator cannot shift the split. A component carrying
 * neither is emitted verbatim, which is what lets an existing key adopt this
 * package without changing its bytes.
 */

import { Sha256 } from "./sha256.js";

/**
 * Total raw size of a component set, in UTF-8 BYTES, above which {@link join}
 * returns a fixed-size hashed identity instead of an escaped join.
 *
 * Bytes, not UTF-16 code units: the Go half measures `len(s)`, so measuring
 * `String.length` here would cross the threshold at a different point for any
 * multibyte input and break parity. The conformance fixture pins a case that
 * fails exactly that way.
 */
export const MAX_COMPONENT_BYTES = 8 << 10;

/** The character {@link join} places between components. */
export const SEPARATOR = ":";

/** The character {@link join} prefixes to a literal separator or escape. */
export const ESCAPE = "\\";

const HASHED_PREFIX = "sha256:";
const HASHED_HEX_LEN = 64;

/** Thrown by {@link split} when a key is a hashed identity, whose components are not recoverable. */
export class HashedKeyError extends Error {
  constructor() {
    super("keyenc: key is a hashed identity and cannot be split");
    this.name = "HashedKeyError";
  }
}

/** Thrown by {@link split} when a key was not produced by {@link join}. */
export class MalformedKeyError extends Error {
  constructor() {
    super("keyenc: key was not produced by join");
    this.name = "MalformedKeyError";
  }
}

const encoder = new TextEncoder();

/**
 * UTF-8 byte length, counted without allocating an encoded copy.
 *
 * Matches `TextEncoder` exactly, including its treatment of a lone surrogate as
 * the 3-byte replacement character — the hashing path encodes through
 * `TextEncoder`, so a counter that disagreed would put the size decision and
 * the digest on different views of the same string.
 */
function utf8Length(s: string): number {
  let n = 0;
  for (const ch of s) {
    // Iteration yields code points, so a well-formed surrogate pair arrives as
    // one two-unit string: 4 UTF-8 bytes. A lone surrogate arrives as one unit
    // and falls through to the 3-byte case, which is what TextEncoder charges
    // for the replacement character it substitutes.
    if (ch.length === 2) {
      n += 4;
      continue;
    }
    const cu = ch.charCodeAt(0);
    if (cu <= 0x7f) {
      n += 1;
    } else if (cu <= 0x7ff) {
      n += 2;
    } else {
      n += 3;
    }
  }
  return n;
}

function escapePart(part: string): string {
  // Escape order matters: doubling backslashes first keeps the mapping
  // injective, because the escapes introduced for separators must not
  // themselves be doubled afterwards. Both reserved characters are ASCII, so
  // iterating code points rather than code units cannot split one.
  let out = "";
  for (const c of part) {
    if (c === ESCAPE || c === SEPARATOR) {
      out += ESCAPE;
    }
    out += c;
  }
  return out;
}

function hashParts(parts: readonly string[]): string {
  const h = new Sha256();
  const lenBuf = new Uint8Array(8);
  const view = new DataView(lenBuf.buffer);
  for (const part of parts) {
    const bytes = encoder.encode(part);
    // 8-byte big-endian length prefix, matching Go's binary.BigEndian.PutUint64.
    // A component longer than 2^32 bytes cannot exist in a JS string, so the
    // high word is always zero.
    view.setUint32(0, 0, false);
    view.setUint32(4, bytes.length, false);
    h.update(lenBuf);
    h.update(bytes);
  }
  return HASHED_PREFIX + h.hex();
}

/**
 * Encode `parts` as one key such that no part's content can forge a different
 * split: `join("a:b", "c")` and `join("a", "b:c")` differ, where a naive
 * template literal makes them identical.
 *
 * Above {@link MAX_COMPONENT_BYTES} of total raw input it returns the set's
 * hashed identity instead. An in-bound set whose escaped join would itself
 * begin with the hashed-identity prefix is also hashed, which keeps the raw and
 * hashed output domains disjoint.
 *
 * No parts encodes as the empty string; a single empty part is hashed, so it
 * cannot alias the zero-part encoding.
 */
export function join(...parts: string[]): string {
  if (parts.length === 0) {
    return "";
  }
  if (parts.length === 1 && parts[0] === "") {
    return hashParts(parts);
  }
  let total = 0;
  for (const part of parts) {
    total += utf8Length(part);
  }
  if (total <= MAX_COMPONENT_BYTES) {
    const joined = parts.map(escapePart).join(SEPARATOR);
    if (!joined.startsWith(HASHED_PREFIX)) {
      return joined;
    }
  }
  return hashParts(parts);
}

/**
 * Recover the exact components {@link join} encoded. Its accepted language is
 * exactly `join`'s image: an escape before anything other than a reserved
 * character, or a trailing dangling escape, throws {@link MalformedKeyError}
 * rather than being normalized, so "split accepted it" is a statement about the
 * key's provenance. A hashed identity throws {@link HashedKeyError}.
 *
 * The empty key splits to no components, mirroring `join()`.
 */
export function split(key: string): string[] {
  if (key === "") {
    return [];
  }
  if (key.startsWith(HASHED_PREFIX)) {
    throw new HashedKeyError();
  }
  const parts: string[] = [];
  let cur = "";
  let escaped = false;
  // Both reserved characters are ASCII, so iterating code points rather than
  // code units cannot split one, and reassembling by concatenation is exact.
  for (const c of key) {
    if (escaped) {
      if (c !== ESCAPE && c !== SEPARATOR) {
        throw new MalformedKeyError();
      }
      cur += c;
      escaped = false;
    } else if (c === ESCAPE) {
      escaped = true;
    } else if (c === SEPARATOR) {
      parts.push(cur);
      cur = "";
    } else {
      cur += c;
    }
  }
  if (escaped) {
    throw new MalformedKeyError();
  }
  parts.push(cur);
  return parts;
}

/**
 * Whether `key` is a hashed identity, i.e. whether {@link join} reduced an
 * oversized component set rather than encoding it. Call it before
 * {@link split} at a site that must parse its keys back.
 */
export function isHashed(key: string): boolean {
  if (!key.startsWith(HASHED_PREFIX)) {
    return false;
  }
  const rest = key.slice(HASHED_PREFIX.length);
  return rest.length === HASHED_HEX_LEN && /^[0-9a-f]+$/.test(rest);
}

File: keyenc.go
package keyenc

import (
	"crypto/sha256"
	"encoding/binary"
	"encoding/hex"
	"errors"
	"strings"
)

// MaxComponentBytes is the total raw size of a component set above which
// [Join] returns a fixed-size hashed identity instead of an escaped join.
//
// The threshold is measured on the RAW components, not on the escaped output,
// so a set's representation never depends on how many separators escaping had
// to double. A delimiter-heavy but honest set whose escaped form is twice this
// bound still encodes as an escaped join; only genuinely oversized input
// crosses over. That keeps the shape of an honest key independent of its
// content, which matters wherever a key is persisted or compared across runs.
const MaxComponentBytes = 8 << 10

// Separator is the character [Join] places between components and [Split]
// reads as a component boundary. It is exported so a caller can assert its own
// field alphabet against the grammar, not so it can be changed.
const Separator = ':'

// Escape is the character [Join] prefixes to a literal [Separator] or [Escape]
// inside a component.
const Escape = '\\'

// hashedPrefix marks a hashed component identity. The raw encodings exclude it
// (see [Join]) so the two output domains cannot collide: without the rule, a
// small component set literally spelling "sha256:<hex>" would alias the hashed
// identity of a different, oversized set. The prefix contains a [Separator] on
// purpose - "sha256" plus a hex digest is a shape an ordinary two-component
// join can produce, so the domains must be separated by construction rather
// than by hoping no caller joins that pair.
const hashedPrefix = "sha256:"

// hashedHexLen is the digest length of the hashed identity, in hex characters.
const hashedHexLen = sha256.Size * 2

var (
	// ErrHashed reports that a key is a hashed identity and therefore carries
	// no recoverable components. Hashing is one-way: the components are gone,
	// not merely encoded. A site that both builds oversized keys and parses
	// them back has to keep the fields it needs somewhere else.
	ErrHashed = errors.New("keyenc: key is a hashed identity and cannot be split")

	// ErrMalformed reports that a key was not produced by [Join]: it ends in a
	// dangling escape, or an escape precedes a character that never needs
	// escaping. Split's accepted language is exactly Join's image, so a
	// tampered or hand-built key is refused rather than silently normalized -
	// which matters wherever a key arrives from a persisted file, a URL
	// segment or client storage.
	ErrMalformed = errors.New("keyenc: key was not produced by Join")
)

// partEscaper escapes the two reserved characters. Escape is listed first so a
// component's literal backslashes are doubled before any separator escape
// introduces new ones, which is what keeps the mapping injective.
var partEscaper = strings.NewReplacer(
	string(Escape), string(Escape)+string(Escape),
	string(Separator), string(Escape)+string(Separator),
)

// Join encodes parts as one key such that no part's content can forge a
// different split: each part is escaped before the separators are inserted, so
// Join("a:b", "c") and Join("a", "b:c") differ, where a naive concatenation
// makes them identical. A part containing neither reserved character is
// emitted verbatim, so a key built from separator-free fields is
// byte-identical to its naive concatenation.
//
// Above [MaxComponentBytes] of total raw input it returns the set's hashed
// identity instead (see [MaxComponentBytes] and [IsHashed]). An in-bound set
// whose escaped join would itself begin with the hashed-identity prefix is
// also routed through the hash, which is what keeps the raw and hashed output
// domains disjoint and therefore keeps Join injective across the size
// boundary.
//
// Two degenerate cases are worth naming because they are the only places the
// output is not simply an escaped join. No parts at all encodes as the empty
// string. A single empty part is routed through the hash, because escaping
// preserves element boundaries only once the join emits a byte: without the
// special case, Join() and Join("") would share the empty encoding and alias
// two distinct component sets.
func Join(parts ...string) string {
	if len(parts) == 0 {
		return ""
	}
	if len(parts) == 1 && parts[0] == "" {
		return hashParts(parts)
	}
	total := 0
	for _, p := range parts {
		total += len(p)
	}
	if total <= MaxComponentBytes {
		if joined := escapeJoin(parts); !strings.HasPrefix(joined, hashedPrefix) {
			return joined
		}
	}
	return hashParts(parts)
}

// Split recovers the exact components [Join] encoded. It is Join's inverse for
// every in-bound key: Split(Join(parts...)) equals parts, and Join(Split(key))
// equals key.
//
// Its accepted language is exactly Join's image. A key carrying an escape
// before anything other than a reserved character, or a trailing dangling
// escape, is refused with [ErrMalformed] rather than normalized, so "Split
// accepted it" is a statement about the key's provenance. A hashed identity is
// refused with [ErrHashed], whose components are not recoverable. The empty key
// splits to no components, mirroring Join's zero-part case.
//
// Split exists so a site that parses its keys back has one grammar rather than
// an encoder plus a separately-maintained parser that can disagree with it.
func Split(key string) ([]string, error) {
	if key == "" {
		return nil, nil
	}
	if strings.HasPrefix(key, hashedPrefix) {
		return nil, ErrHashed
	}
	parts := make([]string, 0, strings.Count(key, string(Separator))+1)
	var cur strings.Builder
	escaped := false
	for i := range len(key) {
		c := key[i]
		switch {
		case escaped:
			if c != Escape && c != Separator {
				return nil, ErrMalformed
			}
			cur.WriteByte(c)
			escaped = false
		case c == Escape:
			escaped = true
		case c == Separator:
			parts = append(parts, cur.String())
			cur.Reset()
		default:
			cur.WriteByte(c)
		}
	}
	if escaped {
		return nil, ErrMalformed
	}
	return append(parts, cur.String()), nil
}

// IsHashed reports whether key is a hashed identity, i.e. whether [Join]
// reduced an oversized component set rather than encoding it. Call it before
// [Split] at a site that must parse its keys back; a key is otherwise
// indistinguishable from a raw one only to a caller that does not look.
func IsHashed(key string) bool {
	rest, ok := strings.CutPrefix(key, hashedPrefix)
	if !ok || len(rest) != hashedHexLen {
		return false
	}
	_, err := hex.DecodeString(rest)
	return err == nil
}

// escapeJoin escapes each part before joining, so element boundaries survive
// in the encoding: a part containing a separator is escaped while the joining
// separators stay raw. Parts free of both reserved characters stay
// byte-identical to their naive concatenation.
func escapeJoin(parts []string) string {
	escaped := make([]string, len(parts))
	for i, p := range parts {
		escaped[i] = partEscaper.Replace(p)
	}
	return strings.Join(escaped, string(Separator))
}

// hashParts streams each component into SHA-256 under a length prefix and
// returns the fixed-size "sha256:<hex>" identity. The length prefix is what
// preserves element boundaries through hashing - ["a", "b"] and ["ab"] hash
// differently - so the hashed regime is as injective as the escaped one. The
// components are never joined into a single allocation, which is the point of
// having a bound at all.
func hashParts(parts []string) string {
	h := sha256.New()
	var lenBuf [8]byte
	for _, p := range parts {
		binary.BigEndian.PutUint64(lenBuf[:], uint64(len(p)))
		_, _ = h.Write(lenBuf[:])
		_, _ = h.Write([]byte(p))
	}
	return hashedPrefix + hex.EncodeToString(h.Sum(nil))
}
